lib/Runtime/Library/PropertyRecordUsageCache.cpp

```diff
@@ -0,0 +1,133 @@
+//-------------------------------------------------------------------------------------------------------
+// Copyright (C) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE.txt file in the project root for full license information.
+//-------------------------------------------------------------------------------------------------------
+#include "RuntimeLibraryPch.h"
+
+namespace Js
+{
+    PropertyRecordUsageCache::PropertyRecordUsageCache() :
+        propertyRecord(nullptr),
+        ldElemInlineCache(nullptr),
+        stElemInlineCache(nullptr),
+        hitRate(0)
+    {
+        // Required due to DEFINE_VTABLE_CTOR on parent, but useless
+    }
+
+    PropertyRecordUsageCache::PropertyRecordUsageCache(StaticType* type, const PropertyRecord* propertyRecord) :
+        propertyRecord(propertyRecord),
+        ldElemInlineCache(nullptr),
+        stElemInlineCache(nullptr),
+        hitRate(0)
+    {
+        // TODO: in future, might be worth putting these inline to avoid extra allocations. PIC copy API needs to be updated to support this though
```

Comment:
Actually, given this comment, the solution might be to allocate PropertyRecordUsageCache dynamically rather than statically? If you do that and then put the inline caches inline on it, you end up net-saving one allocation :)

Reply:
Although, yes, you'd still have to be careful doing that since now your inline caches become interior pointers 

---
In reply to: [162229422](https://github.com/Microsoft/ChakraCore/pull/4569#discussion_r162229422) [](ancestors = 162229422)

Reply:
I'd prefer to leave this alone for now (this comment was just copied over from PropertyString, along with most of this class's implementation).

---
In reply to: [162229510](https://github.com/Microsoft/ChakraCore/pull/4569#discussion_r162229510) [](ancestors = 162229510,162229422)